core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/function/xsd/TestDateCast.java

```diff
@@ -0,0 +1,131 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *******************************************************************************/
+package org.eclipse.rdf4j.query.algebra.evaluation.function.xsd;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.eclipse.rdf4j.model.vocabulary.XSD.DATE;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
+
+import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.datatypes.XMLDatatypeUtil;
+import org.eclipse.rdf4j.model.impl.SimpleValueFactory;
+import org.eclipse.rdf4j.query.algebra.evaluation.ValueExprEvaluationException;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestDateCast {
+
+	private DateCast dateCast;
+	private final ValueFactory vf = SimpleValueFactory.getInstance();
+
+	@Before
+	public void setUp() throws Exception {
+		dateCast = new DateCast();
+	}
+
+	@Test
+	public void testCastPlainLiteral_date() {
+		// Arrange
+		Literal plainLit = vf.createLiteral("1999-09-09");
+		Literal result = null;
+
+		// Act
+		try {
+			result = dateCast.evaluate(vf, plainLit);
+		} catch (ValueExprEvaluationException e) {
+			fail(e.getMessage());
+		}
+
+		// Assert
+		assertNotNull(result);
+		assertThat(result.getLabel()).isEqualTo("1999-09-09");
+		assertThat(result.getDatatype()).isEqualTo(DATE);
+	}
+
+	@Test
+	public void testCastPlainLiteral_date_invalid() {
+		// Arrange
+		Literal plainLit = vf.createLiteral("1999-09-xx");
+
+		// Act & Assert
+		assertThatExceptionOfType(ValueExprEvaluationException.class).isThrownBy(() -> dateCast.evaluate(vf, plainLit));
+	}
+
+	@Test
+	public void testCastPlainLiteral_dateTime() {
+		// Arrange
+		Literal plainLit = vf.createLiteral("1999-09-09T14:45:13");
+		Literal result = null;
+
+		// Act
+		try {
+			result = dateCast.evaluate(vf, plainLit);
+		} catch (ValueExprEvaluationException e) {
+			fail(e.getMessage());
+		}
+
+		// Assert
+		assertNotNull(result);
+		assertThat(result.getLabel()).isEqualTo("1999-09-09T14:45:13");
```

Comment:
@hmottestad Is this result correct or should it be "1999-09-09" ?

Reply:
My understanding is that it should extract the year, month, date and timezone.

<img width="794" alt="Screenshot 2022-09-16 at 21 27 56" src="https://user-images.githubusercontent.com/797185/190715962-49427175-896e-459c-b4c0-3532fae32e99.png">

I think you also need to keep the timezone when casting from xsd:date or from an untyped string.

Reply:
Added the timezone part and added a few more tests.